src/main/java/org/spongepowered/api/text/selector/SpongeSelectorFactory.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered.org <http://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.selector;
+
+import java.util.regex.Pattern;
+
+import org.spongepowered.api.util.annotation.NonnullByDefault;
+import org.spongepowered.mod.text.selector.SpongeSelectorBuilder;
+
+@NonnullByDefault
+public class SpongeSelectorFactory implements SelectorFactory {
+
+    private static final Pattern tokenPattern = Pattern.compile("^@([pare])(?:\\[([\\w=,!-]*)\\])?$");
+
+    @Override
+    public SelectorBuilder createEmptyBuilder() {
+        return new SpongeSelectorBuilder();
+    }
+
+    @Override
+    public Selector parseRawSelector(String selector) {
+        if (tokenPattern.matcher(selector).matches()) {
+            SelectorBuilder builder = createEmptyBuilder().type(SelectorTypes.valueOf(selector.substring(1, 2)).get());
```

Comment:
https://gist.github.com/kenzierocks/80d714dc1910719038c5

`getArgumentMap` is the replacement for your arg-parsing stuff, adapt it to your needs. the `main` function is just a test.